### PR TITLE
Fix for groups synchronization between `etc/shared` folder and wazuh-db to allow the addition of groups with a name starting with `.` or `..` to the `group` table in the global.db database

### PR DESCRIPTION
**Pull request by DProvinciani**
|Related issue|
|---|
| https://github.com/wazuh/wazuh/issues/12742 |

## Description

This pull request implements two modifications to allow the addition of groups with a name starting with `.` or `..` to the `group` table in the global.db database in the **wm_database** synchronization logic:

- The periodical synchronization (executed also during the **modulesd** initialization) has been modified to avoid discharging folders starting with `.` or `..`.
- The synchronization mechanism based on **inotify** was modified to avoid discarding items starting with `.` or `..`.

Other changes:

- In addition, it is implemented a modification to avoid duplicating an array with the names of all the groups. Instead, now it works directly with the data received in a JSON from wazuh-db.

### Adding a group called `test_group`

- Addition

```
root@UbuntuDEV:/home/diego/wazuh-repos/wazuh# /var/ossec/bin/agent_groups -a -g test_group
Do you want to create the group 'test_group'? [y/N]: y
Group 'test_group' created.
```

- Logs

```
2022/03/16 14:46:43 wazuh-modulesd:database[23627] wm_database.c:880 at wm_inotify_start(): DEBUG: inotify: i='0', name='test_group', mask='1073742080', wd='2'
2022/03/16 14:46:43 wazuh-modulesd:database[23627] wm_database.c:941 at wm_inotify_push(): DEBUG: Adding 'etc/shared/test_group' to path table.
2022/03/16 14:46:43 wazuh-modulesd:database[23627] wm_database.c:972 at wm_inotify_pop(): DEBUG: Taking 'etc/shared/test_group' from path table.
2022/03/16 14:46:43 wazuh-modulesd:database[23627] wm_database.c:670 at wm_sync_file(): DEBUG: Synchronizing file 'etc/shared/test_group'
2022/03/16 14:46:43 wazuh-db[23432] wdb_parser.c:670 at wdb_parse(): DEBUG: Global query: find-group test_group
2022/03/16 14:46:43 wazuh-db[23432] wdb_parser.c:670 at wdb_parse(): DEBUG: Global query: insert-agent-group test_group
2022/03/16 14:46:43 wazuh-modulesd:database[23627] wm_database.c:661 at wm_sync_shared_group(): DEBUG: wm_sync_shared_group(): 0.158 ms.
```

### Adding a group called `.test_group`

- Addition

```
root@UbuntuDEV:/home/diego/wazuh-repos/wazuh# /var/ossec/bin/agent_groups -a -g .test_group
Do you want to create the group '.test_group'? [y/N]: y
Group '.test_group' created.
```

- Logs

```
2022/03/16 14:51:05 wazuh-modulesd:database[23627] wm_database.c:880 at wm_inotify_start(): DEBUG: inotify: i='0', name='.test_group', mask='1073742080', wd='2'
2022/03/16 14:51:05 wazuh-modulesd:database[23627] wm_database.c:941 at wm_inotify_push(): DEBUG: Adding 'etc/shared/.test_group' to path table.
2022/03/16 14:51:05 wazuh-modulesd:database[23627] wm_database.c:972 at wm_inotify_pop(): DEBUG: Taking 'etc/shared/.test_group' from path table.
2022/03/16 14:51:05 wazuh-modulesd:database[23627] wm_database.c:670 at wm_sync_file(): DEBUG: Synchronizing file 'etc/shared/.test_group'
2022/03/16 14:51:05 wazuh-db[23432] wdb_parser.c:670 at wdb_parse(): DEBUG: Global query: find-group .test_group
2022/03/16 14:51:05 wazuh-db[23432] wdb_parser.c:670 at wdb_parse(): DEBUG: Global query: insert-agent-group .test_group
2022/03/16 14:51:05 wazuh-modulesd:database[23627] wm_database.c:661 at wm_sync_shared_group(): DEBUG: wm_sync_shared_group(): 0.117 ms.
```

### Adding a group called `..test_group`

- Addition

```
root@UbuntuDEV:/home/diego/wazuh-repos/wazuh# /var/ossec/bin/agent_groups -a -g ..test_group
Do you want to create the group '..test_group'? [y/N]: y
Group '..test_group' created.
```

- Logs

```
2022/03/16 14:51:51 wazuh-modulesd:database[23627] wm_database.c:880 at wm_inotify_start(): DEBUG: inotify: i='0', name='..test_group', mask='1073742080', wd='2'
2022/03/16 14:51:51 wazuh-modulesd:database[23627] wm_database.c:941 at wm_inotify_push(): DEBUG: Adding 'etc/shared/..test_group' to path table.
2022/03/16 14:51:51 wazuh-modulesd:database[23627] wm_database.c:972 at wm_inotify_pop(): DEBUG: Taking 'etc/shared/..test_group' from path table.
2022/03/16 14:51:51 wazuh-modulesd:database[23627] wm_database.c:670 at wm_sync_file(): DEBUG: Synchronizing file 'etc/shared/..test_group'
2022/03/16 14:51:51 wazuh-db[23432] wdb_parser.c:670 at wdb_parse(): DEBUG: Global query: find-group ..test_group
2022/03/16 14:51:51 wazuh-db[23432] wdb_parser.c:670 at wdb_parse(): DEBUG: Global query: insert-agent-group ..test_group
2022/03/16 14:51:51 wazuh-modulesd:database[23627] wm_database.c:661 at wm_sync_shared_group(): DEBUG: wm_sync_shared_group(): 0.087 ms.
```

### Database verification

```
root@UbuntuDEV:/home/diego/wazuh-repos/wazuh# python3 ../../wazuh-tools/qa/wdb-query.py global 'sql SELECT * FROM `group`'
[
    {
        "id": 1,
        "name": "default"
    },
    {
        "id": 2,
        "name": "test_group"
    },
    {
        "id": 3,
        "name": ".test_group"
    },
    {
        "id": 4,
        "name": "..test_group"
    }
]
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation

- Memory tests for Linux
  - [x] Scan-build report
  image

### CODE REVIEW

File: src/wazuh_db/helpers/wdb_global_helpers.c [hunk 1]
```diff
@@ -695,77 +688,67 @@ int wdb_update_groups(const char *dirname, int *sock) {
...
-            n++;
-        }
-
-        item=item->next;
-    }
-
-    array[n] = NULL;
-    cJSON_Delete(root);
+    cJSON *item = NULL;
+    cJSON_ArrayForEach(item, root) {
```
Comment (pereyra-m) [inline src/wazuh_db/helpers/wdb_global_helpers.c:692]: Great improvement !

File: src/wazuh_modules/wm_database.c [hunk 2]
```diff
@@ -890,11 +883,6 @@ static void * wm_inotify_start(__attribute__((unused)) void * args) {
                     mterror(WM_DATABASE_LOGTAG, "Inotify event too large (%u)", event->len);
                     break;
                 }
-
-                if (event->name[0] == '.') {
-                    mtdebug2(WM_DATABASE_LOGTAG, "Discarding hidden file.");
```
Comment (pereyra-m) [inline src/wazuh_modules/wm_database.c:895]: The implications of the hidden files should be discussed with the team

**Review by pereyra-m — CHANGES_REQUESTED** [review]
GJ!

We need to update the `wm_sync_legacy_groups_files()` method also.

File: src/wazuh_modules/wm_database.c (hunk 2 above)
Comment (pereyra-m) [reply src/wazuh_modules/wm_database.c:895]: We should allow only hidden folders and not hidden files to avoid unnecessary **inotify** events.

Comment (DProvinciani) [reply src/wazuh_modules/wm_database.c:895]: As we agreed after analyzing this, the piece of code was restored with the inclusion of a check to only discard hidden files events but handle hidden directory events.

**Review by pereyra-m — APPROVED** [review]
LGTM!

**Pull request merged**